### PR TITLE
pubsub: add conformance test for GCP PubSub driver

### LINKED ISSUE #740: pubsub: add conformance test for GCP driver
This should include a main.tf to create a GCP PubSub topic and subscription. See https://www.terraform.io/docs/providers/google/r/pubsub_topic.html, https://www.terraform.io/docs/providers/google/r/pubsub_subscription.html.

### PR DESCRIPTION
**Pull request by ijt**
Fixes #740

### CODE REVIEW

File: internal/pubsub/gcppubsub/gcppubsub_test.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		return nil, nil, fmt.Errorf("making subscription client: %v", err)
+	}
+	ds, err := openSubscription(ctx, subClient, projectID, subscriptionName)
+	if err != nil {
+		return nil, nil, fmt.Errorf("opening subscription: %v", err)
+	}
+	return dt, ds, nil
+}
+
+func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
```
Comment (jba) [inline internal/pubsub/gcppubsub/gcppubsub_test.go]: Maybe call the `NewXXXClient` methods here, and close the clients in `Close`, to avoid creating new clients on every call to `MakePair`.

File: internal/pubsub/mempubsub/mem.go [hunk 2]
```diff
@@ -183,7 +180,7 @@ func (s *subscription) ReceiveBatch(ctx context.Context) ([]*driver.Message, err
 	// alternative would be complicated by the need to recognize expired messages
 	// promptly.
 	for {
-		if msgs := s.receiveNoWait(time.Now(), maxBatchSize); len(msgs) > 0 {
+		if msgs := s.receiveNoWait(time.Now(), maxMessages); len(msgs) > 0 {
```
Comment (jba) [inline internal/pubsub/mempubsub/mem.go]: You need to fix `receiveNoWait` to honor `maxMessages`. Although I'm not sure why any of this is in this PR.

Comment (ijt) [reply internal/pubsub/mempubsub/mem.go]: You're right, I should move this stuff into a separate PR.

Comment (ijt) [reply internal/pubsub/mempubsub/mem.go]: Moved it to #775 .

File: internal/pubsub/gcppubsub/gcppubsub_test.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	projectID        = "go-cloud-test-216917"
+)
+
+var pathToPrivateKey = flag.String("privatekey", "", "path to .pem file containing private key (required for --record)")
+
+type harness struct {
+}
+
+func (h *harness) MakePair(ctx context.Context) (driver.Topic, driver.Subscription, error) {
+	pubClient, err := raw.NewPublisherClient(ctx)
```
Comment (vangent) [inline internal/pubsub/gcppubsub/gcppubsub_test.go]: You need to make the clients with the record/replay wrapper. See `gcsblob_test.go` for an example using HTTP.

File: internal/pubsub/gcppubsub/gcppubsub_test.go (hunk 1 above)
Comment (vangent) [reply internal/pubsub/gcppubsub/gcppubsub_test.go]: +1

File: internal/pubsub/gcppubsub/gcppubsub_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	serviceAccountID = "pubsub@go-cloud-test-216917.iam.gserviceaccount.com"
+	projectID        = "go-cloud-test-216917"
+)
+
+var pathToPrivateKey = flag.String("privatekey", "", "path to .pem file containing private key (required for --record)")
+
+type harness struct {
+}
+
+func (h *harness) MakePair(ctx context.Context) (driver.Topic, driver.Subscription, error) {
```
Comment (vangent) [inline internal/pubsub/gcppubsub/gcppubsub_test.go]: I think you will end up needing to change this to take some kind of topic/subscription identifiers; currently what you have only supports opening one topic (multiple times) and one subscription (multiple times).

**Comment by vangent** [comment]
@ijt please assign back to reviewers when ready.

**Comment by jba** [comment]
So based on the conversation we just had at the standup, Robert and I are waiting for Issac's PTAL before we look at this PR again. Right?

File: internal/pubsub/gcppubsub/gcppubsub_test.go (hunk 3 above)
Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub_test.go]: Done.

File: internal/pubsub/gcppubsub/gcppubsub_test.go (hunk 4 above)
Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub_test.go]: True, probably also beyond the scope of this PR.

Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub_test.go]: Filed #787.

File: internal/pubsub/gcppubsub/gcppubsub_test.go (hunk 1 above)
Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub_test.go]: Done.

File: internal/pubsub/drivertest/drivertest.go [hunk 5]
```diff
@@ -53,12 +53,6 @@ func RunConformanceTests(t *testing.T, newHarness HarnessMaker) {
 	t.Run("TestSendReceive", func(t *testing.T) {
 		testSendReceive(t, newHarness)
 	})
-	t.Run("TestErrorOnSendToClosedTopic", func(t *testing.T) {
```
Comment (vangent) [inline internal/pubsub/drivertest/drivertest.go]: Why are you deleting tests?

File: internal/pubsub/gcppubsub/gcppubsub.go [hunk 6]
```diff
@@ -51,18 +51,18 @@ func OpenTopic(ctx context.Context, client *raw.PublisherClient, proj gcp.Projec
...
+	ok, err := topicExists(ctx, client, path)
 	if err != nil {
 		return nil, fmt.Errorf("gcppubsub: checking existence of topic: %v", err)
 	}
 	if !ok {
 		return nil, fmt.Errorf("gcppubsub: topic named %q does not exist", topicName)
 	}
-	path := fmt.Sprintf("projects/%s/topics/%s", proj, topicName)
 	return &topic{path, client}, nil
 }
```
Comment (vangent) [inline internal/pubsub/gcppubsub/gcppubsub.go]: What is "_deleted-topic_"?

Comment (vangent) [inline internal/pubsub/gcppubsub/gcppubsub.go]: Should be topicPath now, right? It's not just the name.

**Comment by vangent** [comment]
> I attempted to add record/replay, but it's reporting an error while replaying on Travis:


It looks like it's trying to make an actual RPC. Can you duplicate this running locally without `-record`? Make sure that it's actually using the replay data, not making new RPCs.

File: internal/pubsub/drivertest/drivertest.go [hunk 7]
```diff
@@ -74,6 +74,7 @@ func testSendReceive(t *testing.T, newHarness HarnessMaker) {
 
 	// Send to the topic.
 	ms := []*pubsub.Message{}
+	ss := []string{}
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/pubsub/drivertest/drivertest.go]: Why?

File: internal/pubsub/drivertest/drivertest.go [hunk 8]
```diff
@@ -83,21 +84,27 @@ func testSendReceive(t *testing.T, newHarness HarnessMaker) {
...
 		}
 		ms2 = append(ms2, m2)
+		ss2 = append(ss2, string(m2.Body))
+		m2.Ack()
 	}
 
 	// Check that the received messages match the sent ones.
-	less := func(x, y *pubsub.Message) bool { return bytes.Compare(x.Body, y.Body) < 0 }
-	if diff := cmp.Diff(ms2, ms, cmpopts.SortSlices(less), cmpopts.IgnoreUnexported(pubsub.Message{})); diff != "" {
+	sort.Strings(ss)
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/pubsub/drivertest/drivertest.go]: Comparing the entire messages gives unreadable output since the bodies are byte slices.

File: internal/pubsub/gcppubsub/gcppubsub.go (hunk 6 above)
Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub.go]: It's a special topic name from the GCP PubSub API. The original code is here:

https://github.com/GoogleCloudPlatform/google-cloud-go/blob/977bdf6a60d16cd466ccbfe6c20bfc20ddf923ba/pubsub/topic.go#L276

Subscriptions to deleted topics get their topic name set to this value. 
https://github.com/GoogleCloudPlatform/google-cloud-go/blob/977bdf6a60d16cd466ccbfe6c20bfc20ddf923ba/pubsub/apiv1/publisher_client.go#L297

Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub.go]: What should be topicPath? Are you saying to run s/path/topicPath/?

Comment (vangent) [reply internal/pubsub/gcppubsub/gcppubsub.go]: Please add a comment explaining this and referencing the documentation.

Comment (vangent) [reply internal/pubsub/gcppubsub/gcppubsub.go]: Sorry, I'm referring to line 69, `func topicExists` which takes a `topicName string` argument. IIUC, that is not a topic name, it's a path. It will be easy to mix them up, so be consistent with the names. I don't care about `topicPath` vs `path`.

**Comment by ijt** [comment]
@vangent, I tried running the test without `-record` locally, with the wifi turned off. It hanged, so I hit ctrl-backslash and it gave this stack trace: 

```
google.golang.org/grpc.(*ClientConn).WaitForStateChange(0xc0000efb80, 0xc1cbe0, 0xc0000aa010, 0x3, 0x1)
        /home/issactrotts/src/google.golang.org/grpc/clientconn.go:390 +0x101
google.golang.org/grpc.DialContext(0xc1cbe0, 0xc0000aa010, 0xb5167f, 0x19, 0xc0000b09c0, 0x4, 0x6, 0x0, 0x
0, 0x0)
        /home/issactrotts/src/google.golang.org/grpc/clientconn.go:298 +0x7db
github.com/google/go-cloud/internal/testing/setup.NewGCPgRPCConn(0xc1cbe0, 0xc0000aa010, 0xc0002da200, 0xb
5167f, 0x19, 0xf, 0x10147ca)
        /home/issactrotts/src/github.com/google/go-cloud/internal/testing/setup/setup.go:131 +0x1b1
github.com/google/go-cloud/internal/pubsub/gcppubsub.newHarness(0xc1cbe0, 0xc0000aa010, 0xc0002da200, 0x1,
 0xf, 0xc0000aaca0, 0xc0000aaca8)
        /home/issactrotts/src/github.com/google/go-cloud/internal/pubsub/gcppubsub/gcppubsub_test.go:56 +0
x6f
github.com/google/go-cloud/internal/pubsub/drivertest.testSendReceive(0xc0002da200, 0xb6c090)
        /home/issactrotts/src/github.com/google/go-cloud/internal/pubsub/drivertest/drivertest.go:64 +0x6b
github.com/google/go-cloud/internal/pubsub/drivertest.RunConformanceTests.func1(0xc0002da200)
```

It hangs in `grpc.DialContext`, but it shouldn't be calling that at all in replay mode.

@jba sent a link to some code he wrote to do replaying with grpc: https://code.googlesource.com/gocloud/+/master/datastore/integration_test.go#136.

I'm going to see if that can be used here.

File: internal/pubsub/drivertest/drivertest.go (hunk 5 above)
Comment (ijt) [reply internal/pubsub/drivertest/drivertest.go]: I think jba is planning to getting rid of the `Close` methods, so I don't think it makes sense to keep tests that involve them. But maybe we want to wait until we actually remove those.

File: internal/pubsub/drivertest/drivertest.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ijt) [reply internal/pubsub/drivertest/drivertest.go]: I thought I saw you doing it this way somewhere, but okay. Changing it.

File: internal/pubsub/drivertest/drivertest.go (hunk 5 above)
Comment (ijt) [reply internal/pubsub/drivertest/drivertest.go]: I'll try bringing them back but I think there was a reason why I removed them.

Comment (ijt) [reply internal/pubsub/drivertest/drivertest.go]: Done.

File: internal/pubsub/gcppubsub/gcppubsub.go (hunk 6 above)
Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub.go]: Done.

Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub.go]: Done.

**Comment by ijt** [comment]
PTAL

**Comment by jba** [comment]
How did you fix the replay-when-disconnected issue?

**Comment by vangent** [comment]
Tests are still failing....

**Comment by ijt** [comment]
@jba, the replay-when-disconnected issue isn't fixed but I don't think it's a high priority to fix it.

Ross just found the reason why the tests were failing on Travis: `gcppubsub.NewTopic` and `NewSubscription` were creating clients that shadowed the ones being passed in as parameters, and these shadowing clients made their own connections that didn't do record/replay.

**Comment by ijt** [comment]
Travis failed on Windows: "/c/Users/travis/.travis/job_stages: line 104: go: command not found"

I guess it's a bug in Travis support for windows, nothing to do with this pull request. Some evidence from the top of the Travis page: "Windows builds are in early access stage".

**Comment by vangent** [comment]
> Travis failed on Windows
Travis is intermittently flaky across all builds unfortunately. Just kick it off again (I've done that for your Windows build) to be sure it's flake.

File: internal/pubsub/gcppubsub/gcppubsub_test.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	//    "Enable API", "Create a topic".
+	// 1b. Create a subscription by clicking on the topic, then clicking on
+	//    the icon at the top with a "Create subscription" tooltip.
+	// 2. Update the topicName constant to your topic name, and the
+	//    subscriptionName to your subscription name.
+	// 3. Create a service account in your GCP project and update the
+	//    serviceAccountID constant to it.
+	topicName        = "test-topic"
+	subscriptionName = "test-subscription-1"
+	serviceAccountID = "pubsub@go-cloud-test-216917.iam.gserviceaccount.com"
```
Comment (vangent) [inline internal/pubsub/gcppubsub/gcppubsub_test.go]: This constant `serviceAccountID` is never used...?

File: internal/pubsub/gcppubsub/gcppubsub_test.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+)
+
+type harness struct {
+	closer    func()
+	pubClient *raw.PublisherClient
+	subClient *raw.SubscriberClient
+}
+
+func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
+	endPoint := "pubsub.googleapis.com:443"
```
Comment (vangent) [inline internal/pubsub/gcppubsub/gcppubsub_test.go]: The endpoint should probably go into a helper in `gcppubsub`, similar to what's in [runtimevar](https://github.com/google/go-cloud/blob/master/runtimevar/runtimeconfigurator/runtimeconfigurator.go#L56). OK to do in a separate PR.

File: internal/pubsub/gcppubsub/gcppubsub_test.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+func (h *harness) MakeTopic(ctx context.Context) (driver.Topic, error) {
+	dt, err := openTopic(ctx, h.pubClient, projectID, topicName)
+	if err != nil {
+		return nil, fmt.Errorf("opening topic: %v", err)
+	}
+	return dt, nil
+}
+
+func (h *harness) MakeSubscription(ctx context.Context, dt driver.Topic) (driver.Subscription, error) {
```
Comment (vangent) [inline internal/pubsub/gcppubsub/gcppubsub_test.go]: Per meatspace discussion, I don't see why `MakeSubscription` takes a `dt`. File an issue?

File: internal/pubsub/gcppubsub/gcppubsub_test.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+	ds, err := openSubscription(ctx, h.subClient, projectID, subscriptionName)
+	if err != nil {
+		return nil, fmt.Errorf("opening subscription: %v", err)
+	}
+	return ds, nil
+}
+
+func (h *harness) Close() {
+	h.closer()
```
Comment (vangent) [inline internal/pubsub/gcppubsub/gcppubsub_test.go]: I think it makes more sense to close `h` after the clients, as this is the underlying transport for `pubClient` and `subClient`.

File: internal/pubsub/gcppubsub/gcppubsub_test.go (hunk 10 above)
Comment (ijt) [reply internal/pubsub/gcppubsub/gcppubsub_test.go]: Filed https://github.com/google/go-cloud/issues/857.

**Comment by ijt** [comment]
@vangent, PTAL.

File: internal/pubsub/gcppubsub/gcppubsub_test.go (hunk 11 above)
Comment (vangent) [reply internal/pubsub/gcppubsub/gcppubsub_test.go]: It should work by topic name.
